### ISSUE TITLE
networkd 249 with ManageForeignRoutes=off fails to set up its own static routes

### ISSUE THREAD

**Issue by grawity** [systemd/systemd#20201]
After upgrading systemd to v249 on a few VMs, they were all stuck on "waiting for network to be configured" on the next boot. When inspecting through debug-shell, all of them had IP addresses configured – but the default routes were missing entirely.

<sup>_(The previously mentioned crash was due to a distro issue, it is already fixed in 248-stable.)_</sup>

The hosts use just regular systemd-networkd with static addresses and gateways:

```
[Match]
Name=eth0

[Network]
Address=2.59.135.246/24
Gateway=2.59.135.1
Address=2a0d:5941:1:bb::2/48
Gateway=2a0d:5941:1:0::1
IPv6AcceptRA=no
```

However, the global networkd.conf has ManageForeignRoutes disabled (because some of the VMs have fairly huge non-main routing tables, which used to severaly impact networkd's resource usage):

```
# /etc/systemd/networkd.conf.d/foreign-routes.conf
[Network]
ManageForeignRoutes=no
```

**systemd version the issue has been seen with**
 > v249

**Used distribution**
 > Arch

**Linux kernel version used** (`uname -a`)
 > 5.10.48-1-lts

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > Networkd should configure static routes from the `[Network] Gateway=` parameter.

**Unexpected behaviour you saw**
 > Addresses and rules are configured, but routes are not.

**Steps to reproduce the problem**
 > …

**Additional program output to the terminal or log subsystem illustrating the issue**

```
sd-device-monitor: No permission to stat PID1's netns, unable to determine if we are in host netns: Operation not permitted
Bus bus-api-network: changing state UNSET → OPENING
sd-bus: starting bus bus-api-network by connecting to /run/dbus/system_bus_socket...
Bus bus-api-network: changing state OPENING → AUTHENTICATING
Registering bus object implementation for path=/org/freedesktop/LogControl1 iface=org.freedesktop.LogControl1
Loaded timestamp for '/etc/systemd/network'.
Failed to read $container of PID 1, ignoring: Operation not permitted
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Found container virtualization none.
Virtualization KVM found in DMI (/sys/class/dmi/id/product_name)
UML virtualization not found in /proc/cpuinfo.
Virtualization found, CPUID=KVMKVMKVM
Found VM virtualization kvm
/usr/lib/systemd/network/80-container-host0.network: Conditions in the file do not match the system environment, skipping.
eth0: Link 2 added
eth0: Saved original MTU 1500 (min: 68, max: 65535)
eth0: Saved hardware address: 00:16:3c:98:50:24
eth0: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
eth0: Link UP
eth0: Gained carrier
eth0: udev initialized link
eth0: State changed: pending -> initialized
lo: Failed to get driver, continuing without: Operation not supported
lo: Link 1 added
lo: Saved original MTU 65536 (min: 0, max: 4294967295)
lo: Saved hardware address: 00:00:00:00:00:00
lo: Flags change: +LOOPBACK +UP +LOWER_UP +RUNNING
lo: Link UP
lo: Gained carrier
lo: udev initialized link
lo: State changed: pending -> initialized
eth0: Remembering foreign address: fe80::216:3cff:fe98:5024/64 (valid forever, preferred forever), flags: permanent
eth0: link_check_ready(): link is in initialized state.
eth0: Gained IPv6LL
eth0: link_check_ready(): link is in initialized state.
eth0: Remembering foreign address: 2a0d:5941:1:bb::2/48 (valid forever, preferred forever), flags: permanent
eth0: link_check_ready(): link is in initialized state.
lo: Remembering foreign address: ::1/128 (valid forever, preferred forever), flags: permanent
lo: link_check_ready(): link is in initialized state.
eth0: Remembering foreign address: 2.59.135.246/24 (valid forever, preferred forever), flags: permanent
eth0: link_check_ready(): link is in initialized state.
lo: Remembering foreign address: 127.0.0.1/8 (valid forever, preferred forever), flags: permanent
lo: link_check_ready(): link is in initialized state.
rtnl: received non-static neighbor, ignoring.
Ignoring received foreign routing policy rule: priority: 32766, ::/0 -> ::/0, iif: n/a, oif: n/a, table: main(254)
Ignoring received foreign routing policy rule: priority: 0, ::/0 -> ::/0, iif: n/a, oif: n/a, table: local(255)
rtnl: received rule message with invalid family 129, ignoring.
rtnl: received rule message with invalid family 128, ignoring.
Ignoring received foreign routing policy rule: priority: 32767, 0.0.0.0/0 -> 0.0.0.0/0, iif: n/a, oif: n/a, table: default(253)
Ignoring received foreign routing policy rule: priority: 32766, 0.0.0.0/0 -> 0.0.0.0/0, iif: n/a, oif: n/a, table: main(254)
Ignoring received foreign routing policy rule: priority: 0, 0.0.0.0/0 -> 0.0.0.0/0, iif: n/a, oif: n/a, table: local(255)
Enumeration completed
eth0: Link state is up-to-date
eth0: found matching network '/etc/systemd/network/10-eth.network'.
eth0: State changed: initialized -> configuring
Setting '/proc/sys/net/ipv6/conf/eth0/disable_ipv6' to '0'
Setting '/proc/sys/net/ipv6/conf/eth0/use_tempaddr' to '0'
Setting '/proc/sys/net/ipv6/conf/eth0/accept_ra' to '0'
Setting '/proc/sys/net/ipv6/conf/eth0/proxy_ndp' to '0'
Setting '/proc/sys/net/ipv4/conf/eth0/promote_secondaries' to '1'
eth0: Requested to set IPv6LL address generation mode
eth0: Requested to set master interface
eth0: Requested to activate link
eth0: LLDP: Started LLDP client
eth0: Started LLDP.
eth0: Requesting address: 2.59.135.246/24 (valid forever, preferred forever), flags: n/a
eth0: Requesting address: 2a0d:5941:1:bb::2/48 (valid forever, preferred forever), flags: n/a
eth0: Setting addresses
eth0: link_check_ready(): link layer is configuring.
eth0: link_check_ready(): link layer is configuring.
eth0: link_check_ready(): link layer is configuring.
eth0: link_check_ready(): link layer is configuring.
eth0: link_check_ready(): link layer is configuring.
eth0: link_check_ready(): link layer is configuring.
eth0: Requesting route: dst: n/a, src: n/a, gw: 2a0d:5941:1::1, prefsrc: n/a, scope: global, table: main(254), proto: static, type: unicast, nexthop: 0, priority: 1024
eth0: Requesting route: dst: n/a, src: n/a, gw: 2.59.135.1, prefsrc: n/a, scope: global, table: main(254), proto: static, type: unicast, nexthop: 0, priority: 0
eth0: Requesting routes
eth0: link_check_ready(): link layer is configuring.
eth0: Setting IPv6LL address generation mode
eth0: Setting master interface
Bus bus-api-network: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RequestName cookie=2 reply_cookie=0 signature=su error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=3 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.19 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus bus-api-network: changing state HELLO → RUNNING
lo: Link state is up-to-date
lo: State changed: initialized -> unmanaged
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_31 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=4 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
eth0: IPv6LL address generation mode set.
eth0: master interface set.
eth0: link_check_ready(): link is not activated.
eth0: Bringing link up
eth0: link_check_ready(): static addresses are not configured.
eth0: Configuring address: 2.59.135.246/24 (valid forever, preferred forever), flags: permanent
eth0: Configuring address: 2a0d:5941:1:bb::2/48 (valid forever, preferred forever), flags: permanent
eth0: Remembering updated address: 2.59.135.246/24 (valid forever, preferred forever), flags: permanent
eth0: link_check_ready(): static addresses are not configured.
eth0: Remembering updated address: 2a0d:5941:1:bb::2/48 (valid forever, preferred forever), flags: permanent
eth0: link_check_ready(): static addresses are not configured.
eth0: Addresses set
eth0: link_check_ready(): static routes are not configured.
Got message type=signal sender=org.freedesktop.DBus.Local destination=n/a path=/org/freedesktop/DBus/Local interface=org.freedesktop.DBus.Local member=Connected cookie=4294967295 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.DBus destination=:1.19 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.DBus destination=:1.19 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=3 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.19 path=n/a interface=n/a member=n/a cookie=4 reply_cookie=2 signature=u error-name=n/a error-message=n/a
Successfully acquired requested service name.
Got message type=method_return sender=org.freedesktop.DBus destination=:1.19 path=n/a interface=n/a member=n/a cookie=5 reply_cookie=3 signature=n/a error-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.login1',path='/org/freedesktop/login1',interface='org.freedesktop.login1.Manager',member='PrepareForSleep' successfully installed.
rtnl: received non-static neighbor, ignoring.
```

**Comment by grawity** [comment]
(I guess the ManageForeignRoutes option might not be necessary anymore, though, so I'll disable it for now and see how it goes.)

**Comment by yuwata** [comment]
Thank you for the report. Fix is waiting in #20202. If possible, please test the PR.

**Comment by grawity** [comment]
Thanks, tested and it seems to be working.
